File: app/src/services/auth.service.ts
import { api } from './api';
import { tokenStorage } from './tokenStorage';
import type { AuthResponse, RegisterRequest, LoginRequest, Profile } from '@ai-companions/shared';

export const authService = {
  async register(data: RegisterRequest): Promise<AuthResponse> {
    const response = await api.post('/auth/register', data);
    const result = response.data.data as AuthResponse;
    await this.saveTokens(result.access_token, result.refresh_token);
    return result;
  },

  async login(data: LoginRequest): Promise<AuthResponse> {
    const response = await api.post('/auth/login', data);
    const result = response.data.data as AuthResponse;
    await this.saveTokens(result.access_token, result.refresh_token);
    return result;
  },

  async logout(): Promise<void> {
    await tokenStorage.removeItem('access_token');
    await tokenStorage.removeItem('refresh_token');
  },

  async forgotPassword(email: string): Promise<void> {
    await api.post('/auth/forgot-password', { email });
  },

  async getProfile(): Promise<Profile> {
    const response = await api.get('/user/profile');
    return response.data.data as Profile;
  },

  async saveTokens(accessToken: string, refreshToken: string): Promise<void> {
    await tokenStorage.setItem('access_token', accessToken);
    await tokenStorage.setItem('refresh_token', refreshToken);
  },

  async hasToken(): Promise<boolean> {
    const token = await tokenStorage.getItem('access_token');
    return !!token;
  },
};

File: ai_companions_flutter/lib/core/services/auth_service.dart
// Authentication Service

import 'package:dio/dio.dart';
import '../../core/network/dio_client.dart';
import '../../core/storage/token_storage.dart';
import '../../shared/models/index.dart';
import 'i_auth_service.dart';

class AuthService implements IAuthService {
  AuthService._();

  static final instance = AuthService._();

  Dio get _dio => DioClient.instance.dio;

  @override
  Future<AuthResponse> register({
    required String email,
    required String password,
    required String username,
    String? displayName,
  }) async {
    try {
      final response = await _dio.post(
        '/auth/register',
        data: {
          'email': email,
          'password': password,
          'username': username,
          'display_name': displayName,
        },
      );

      final data = response.data['data'] as Map<String, dynamic>;
      final authResponse = AuthResponse.fromJson(data);
      await _saveTokens(authResponse.accessToken, authResponse.refreshToken);
      return authResponse;
    } catch (e) {
      rethrow;
    }
  }

  @override
  Future<AuthResponse> login({
    required String email,
    required String password,
  }) async {
    try {
      final response = await _dio.post(
        '/auth/login',
        data: {'email': email, 'password': password},
      );

      final data = response.data['data'] as Map<String, dynamic>;
      final authResponse = AuthResponse.fromJson(data);
      await _saveTokens(authResponse.accessToken, authResponse.refreshToken);
      return authResponse;
    } catch (e) {
      rethrow;
    }
  }

  @override
  Future<void> logout() async {
    await TokenStorage.instance.clearTokens();
  }

  @override
  Future<void> forgotPassword(String email) async {
    try {
      await _dio.post('/auth/forgot-password', data: {'email': email});
    } catch (e) {
      rethrow;
    }
  }

  @override
  Future<Profile> getProfile() async {
    try {
      final response = await _dio.get('/user/profile');
      final data = response.data['data'] as Map<String, dynamic>;
      return Profile.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  @override
  Future<Profile> updateProfile({String? displayName, String? bio}) async {
    try {
      final response = await _dio.put(
        '/user/profile',
        data: {'display_name': displayName, 'bio': bio},
      );
      final data = response.data['data'] as Map<String, dynamic>;
      return Profile.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  @override
  Future<bool> hasToken() async {
    final token = await TokenStorage.instance.getAccessToken();
    return token != null;
  }

  Future<void> _saveTokens(String accessToken, String refreshToken) async {
    await TokenStorage.instance.setAccessToken(accessToken);
    await TokenStorage.instance.setRefreshToken(refreshToken);
  }
}
